### PR TITLE
[RLlib] Fix PyTorch grad cliping usage

### LINKED ISSUE #4965: [RLlib] PyTorch grad_clip happens in place - Extra Grad Info Reports Unclipped Total Norm
@ekl

https://github.com/ray-project/ray/blob/master/python/ray/rllib/agents/a3c/a3c_torch_policy.py#L63

https://pytorch.org/docs/master/_modules/torch/nn/utils/clip_grad.html#clip_grad_norm

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Why are these changes needed?

PyTorch's grad clipping clips in place and returns the global norm before clipping (instead of after clipping).
This throws our gnorm metric off.

## Related issue number

Closes #4965

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by ArturNiederfahrenhorst** [comment]
After fix: https://tensorboard.dev/experiment/bFeobgABTW6IAVl4iBM6lQ/#scalars&tagFilter=gnorm

**Comment by ArturNiederfahrenhorst** [comment]
Before fix: https://tensorboard.dev/experiment/D27Ao1RXSYaT4dLYB4AjZA/#scalars&tagFilter=gnorm

**Comment by ArturNiederfahrenhorst** [comment]
Since it was not using this utility function, QMix went under the radar here.

This PR now also deals with unification of our grad clipping algorithm parameters (grad_clip vs grad_clip_norm) in case of QMix. Offline Estimators do not obey this atm.
QMix before: https://tensorboard.dev/experiment/W34kbWQLTDOrhkAvgM863A/#scalars&tagFilter=grad&_smoothingWeight=0
QMix after: https://tensorboard.dev/experiment/yHo7dcrPQgm0r3R6WdF7Sg/#scalars&tagFilter=grad&_smoothingWeight=0

File: rllib/algorithms/qmix/qmix.py [hunk 1]
```diff
@@ -70,7 +71,7 @@ def __init__(self):
         self.double_q = True
         self.optim_alpha = 0.99
         self.optim_eps = 0.00001
-        self.grad_norm_clipping = 10
+        self.grad_clip = 10
```
Comment (sven1977) [inline rllib/algorithms/qmix/qmix.py:74]: Nice to have this unified!

File: rllib/algorithms/qmix/qmix.py [hunk 2]
```diff
@@ -162,15 +164,29 @@ def training(
...
                 this value.
+            grad_norm_clipping: Depcrecated in favor of grad_clip
 
         Returns:
             This updated TrainerConfig object.
         """
         # Pass kwargs onto super's `training()` method.
         super().training(**kwargs)
 
+        if grad_norm_clipping is not None:
```
Comment (sven1977) [inline rllib/algorithms/qmix/qmix.py:177]: Cool.

File: rllib/algorithms/qmix/qmix_policy.py [hunk 3]
```diff
@@ -266,7 +267,7 @@ def __init__(self, obs_space, action_space, config):
         )
         from torch.optim import RMSprop
 
-        self.optimiser = RMSprop(
+        self.rmsprop_optimizer = RMSprop(
```
Comment (sven1977) [inline rllib/algorithms/qmix/qmix_policy.py:270]: :) "optimiser"

File: rllib/utils/torch_utils.py [hunk 4]
```diff
@@ -51,12 +51,15 @@ def apply_grad_clipping(
             # clip_grad_norm_. Would fail otherwise.
             params = list(filter(lambda p: p.grad is not None, param_group["params"]))
             if params:
-                grad_gnorm = nn.utils.clip_grad_norm_(
-                    params, policy.config["grad_clip"]
-                )
-                if isinstance(grad_gnorm, torch.Tensor):
-                    grad_gnorm = grad_gnorm.cpu().numpy()
-                info["grad_gnorm"] = grad_gnorm
+                # PyTorch clips gradients inplace and returns the norm before clipping
```
Comment (sven1977) [inline rllib/utils/torch_utils.py:54]: Sorry, stupid question: w/o checking. Does QMix Policy actually use `apply_grad_clipping`?

**Review by sven1977 — APPROVED** [review]
Looks good. Just the question on making sure QMix already uses the grad clipping utility. Seems weird that it wouldn't throw an error due to the missing `grad_clip` config key.

File: rllib/utils/torch_utils.py [hunk 5]
```diff
@@ -51,12 +51,15 @@ def apply_grad_clipping(
...
+                # We therefore need to compute grad_gnorm further down (fixes #4965)
+                clip_value = policy.config["grad_clip"]
+                global_norm = nn.utils.clip_grad_norm_(params, clip_value)
+
+                if isinstance(global_norm, torch.Tensor):
+                    global_norm = global_norm.cpu().numpy()
```
Comment (gjoliver) [inline rllib/utils/torch_utils.py:60]: I am curious, can we do this clipping on whatever device it is, like torch.min(), without unloading it from GPU?

File: rllib/utils/torch_utils.py [hunk 6]
```diff
@@ -51,12 +51,15 @@ def apply_grad_clipping(
...
+
+                info["grad_gnorm"] = min(global_norm, clip_value)
```
Comment (gjoliver) [inline rllib/utils/torch_utils.py:62]: I am clueless, will there be negative global_norm?

**Review by gjoliver — COMMENTED** [review]
thanks for fixing this!! sorry I dropped the ball, I was looking at this a bit yesterday.

File: rllib/utils/torch_utils.py (hunk 6 above)
Comment (ArturNiederfahrenhorst) [reply rllib/utils/torch_utils.py:62]: global norm is an l2 norm so it can not become negative!

File: rllib/utils/torch_utils.py (hunk 5 above)
Comment (ArturNiederfahrenhorst) [reply rllib/utils/torch_utils.py:60]: Do you mean that as in "does this really work what we are doing here?".
Because we are clipping on whatever device params are and we unload afterward. And the metrics look very much like it is working.
